utils/docs/gmake-transistors-sym.1.in
.TH gmake-transistors-sym 1 "@DATE@" "gEDA Project" @VERSION@

.SH NAME
gmake-transistors-sym \- Make gEDA Transitor Symbols

.SH SYNOPSIS
.B gmake-transistors-sym
.I symbol_file

.SH SYNOPSIS
.B gmake-transistors-sym
.RI [ OPTIONS ] TEMPLATE[.sym] PARAMETERS[.spm]

.SH DESCRIPTION
.B gmake-transistors-sym
Make gEDA Transitor Symbols

gmake-transistors-sym is for 'batch' production of transistor symbols for the geda-gaf project. The script requires a template symbol file and a parameter file. Parameters read from the file are used in the output symbol file name or used as replacement values for pre-set variables in the symbol template file.

.SH QUICK
.B usage: [-h, --help] | [--version] | [options] template[.sym] parameters[.spm]
example:
gmake-transistors-sym npn-example-template.sym params-exported-gnumeric.spm<ENTER>.
.SH PARAMETERS
Fields in the TAB or space delimited parameter file are read and interpreted as follows:

1. Part Number string to be substituted for PART_NUMBER in the symbol file
2. Voltage string is included in the file name and as part of the COMPONENT_VALUE
3. Features string also included in the file name and in the COMPONENT_VALUE. This  could be, for example, Wattage or Ampacity
4. Package name,  if the -p option is present on the command line then the 4th field interpreted  as the package ID and included in the file name
5. FootPrint must be the next argument and is substituted for unknown in the template
6. Optional symbol file modifier, if present, the value is used as a second suffix before dash suffix. Normally this would be used to select the template with the correct pin -out, for example using ECB or BEC. The string is prefixed with a underscore composing the template file name.
7. Optional symbol file suffix, if this value is present in the parameter file, the string will be used between the template given on the commanline and the symbol file extension, in this case the name of the template given on the command line should only be the base file name. The string is prefixed with a dash composing the template file name. Note that if a modifier is not given then parameter ini position 6 is assumed to be the symbol file suffix

.SH INFORMATION
Use of the following options displays information related to gmake-transistors-sym
but does not create any symbols:
.IP "-h, --help"
Show information about the advanced usage of this command.
.IP "-u, --usage"
Displays examples on usage"
.IP "--version"
Displays software version information"

.SH DIRECTIVES
.IP "-v, --verbose"
Include additional information during processing.

.SH OPTIONS
.IP "-p, --package"
The packge type is included in the paratmeter data
.IP "-t, --template"
Include the template prefix and suffix in the output symbol name
.IP "-m, --modifier"
Suffix the template modifier in the output symbol name, like _ecb

.SH HOMEPAGE
http://www.geda-project.org

.SH BUGS
You can report bugs at the Launchpad site of gmake-transistors-sym:
https://bugs.launchpad.net/geda/gmake-transistors-sym/+filebug

.SH AUTHOR
Wiley Edward Hill < wileyhill dot gmail dot com >

.SH SEE ALSO
.BR gmake-caps-sym (1) gmake-set-resistors (1)

.SH COPYRIGHT
.nf
Copyright \(co 2012-@YEAR@ Wiley Edward Hill.  License GPLv2+: GNU GPL
version 2 or later.  Please see the `COPYING' file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

.


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2017 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf gEDA 2.1.1.1 - 04/01/19 Rampage Edition

* Expanded support for internationalization (i18n).

* All utilities now support internationalization (i18n).

* The visibility of toolbar tooltips can now be control using menu options.

* The context menu available when creating Polygon objects now includes
  an UNDO option allowing users to undo the previous section and resume
  path mode from the previous node.

* Scripting to support printing schematic documents from the command-line
  using gschem has been enhanced to support postscript or pdf output.
  Previously the default was changed from postscript to pdf. Postscript
  outputs could still be produced using menu options available in gschem.
  The output type is now automatically detected when using the command
  line interface by the file extension of the output file name. The
  default remains pdf if the output name is not given.

* Configuration function output-type has been renamed output-type.

* Libgedacairo now uses round end-caps when drawing buses, resulting in
  smoother transitions and improved appearance.

* A new keyword option hide-columns has been added to gattrib to allow
  user to designate a list attributes to be hidden in the GUI be default.
  A new Column Visibility dialog has been added to support user management
  of visible columns.

* A new path backend has been added to the gaf tool to allow scripting
  lanauages other than scheme to retrieve path information from libgeda.

* Improved appearence of various dialogs in gschem.

* gschem now logs exiting of the schematic editor if log-time is enabled.

* gschem now reports failure to attach attributes that aleady attached.

* gschem text properties dialog now supports operations to manipulate
  text case.

* gschem translate dialog now allows for translation of only selected
  objects. Additionally, the translate offset entry now accepts X and
  Y coordinates.

* Auxiliary buffers can now be explicitly cleared using a right context
  menu on the corresponding paste menu item.

* The suite now include template documents install to the user's template
  folder so the user can create new schematic and symbol documents from
  the system's context menu.

* Issues resolved:

* Libgedacairo has been enhanced to resolve issues with the end points of
  arcs shifting at different zoom levels. The shifting sometimes resulted
  in misalignment to adjoining lines at higher magnification levels.

* Extensive revisions were made to the Rampage Edition to build the suite
  on MinGW platforms. The Rampage Edition is now fully functional under
  native Win32 host systems including i18n support.

* The Rename Session dialog now fills in the previous session name,
  previously to dialog presented the user with an empty name field.

* The locked bit is not preserved when copying components in gschem.

* gschem now restarts text search when users change text in the Find Text
  dialog and click "find" or hits ENTER instead of closing the dialog.

* gschem now raises the main window when interacting with the Array Dialog.
  Previously, user could selected other application windows and return to
  the dialog and select the button that lowered the dialog but did not
  ensure the main window was in the foreground.

* An issue was resolved with controls not being initially enabled on the
  Attributes tab of the Preferences Dialog if the Attributes tab was the
  "saved" initial tab in gschem. Switching tabs or the view list controls
  re-enabled the disabled controls. The saved tab is now restored after
  settings are uploaded to the Preferences Dialog and this is reflected
  in the initial state of controls on the Attributes tab.

* An issue was resolved with resetting component-dialog-attributes when
  a second list was encountered in rc files. If a second attributes list
  was processed gschem would fault when the list was accessed or exiting
  gschem.

* The drc2 back end no longer reports duplicate slots for on components
  that do not have slots. This generally occurred when components have
  duplicate reference designators because these would typically show up
  using the same slot number.

* Libgeda no longer reports an oddity regarding symbol versions if the
  oddity was due to the symbol file being missing.

* The scheme file geda-deprecated-config.scm is no longer installed.

Notable changes in gEDA/gaf gEDA 2.1.1.1 - 11/12/17 Rampage Edition

* Expanded support for internationalization (i18n).

* The build system now support using the --with-included-gettext option to
  configure the suite to build and use an in-source international library.

* Font-sizes for widgets on GedaToolbars can now be specified in RC files.

* The toolbar layout in gschem has been improved to better support labels.

* The visibility of toolbar tooltips can now be control using RC keywords.

* The Slot Editor Dialog now incorporates a spinner control to adjust the
  slot number of slotted components. The range is restricted to valid slots
  for the selected component.

* New bitmaps have been added;

      * geda-deslect.png
      * geda-find-attribute
      * geda-properties.png
      * geda-trash-can.png

* libgedathon no longer over-writes existing gafrc when the declare_local_sym
  feature is used.

* libgedathon now supports internationalization (i18n).

Notable changes in gEDA/gaf gEDA 2.1.1.0 - 01/01/17 Rampage Edition

* Improvements were made to both the build system and source code to ensure
  gEDA-GAF builds more reliably on 64 bit platforms. Testing was performed
  using Ubuntu 12.04 and Ubuntu 14.04 64 bit host systems.


utils/gmake/gmake-transistors-sym.sh
#!/bin/bash

VER=1.5

# Author: Wiley Edward Hill
# Date: 03/08/13
#

shopt -s nullglob

MINARGS=1          # Script requires at least one arguments.
ERR_FILE_NOT_FOUND=2
ERR_BAD_ARGS=65

VERBOSE=false
QUIET=false

PrefixLength=3
Extension=".sym"

TemplateFile=""
ParameterFile=""
Prefix=4

include_package=false
include_template=false
include_modifier=false

do_show_usage(){
   echo Usage:   `basename $0` '[-h, --help] | [--version] | [options] template[.sym] parameters[.spm]'
}

do_show_help(){
echo
echo Usage:   `basename $0` '[-options] template-symbol parameter-file'
echo
echo "Abstract: This script is for 'batch' production of transistor symbols"
echo "         for the geda-gaf project. The script uses template symbol file"
echo "         and a parameter file. Parameters read from the file are used"
echo "         in the output symbol file name or used as replacement values for"
echo "         pre-set variables in the symbol template file. The parameters are"
echo "         read and interpreted as follows:"
echo
echo "         1. Part Number string to be substituted for PART_NUMBER in the symbol file"
echo "         2. Voltage string is included in the file name and as part of the COMPONENT_VALUE"
echo "         3. Features string also included in the file name and in the COMPONENT_VALUE. This"
echo "            could be, for example, Wattage or Ampacity"
echo "         4. Package name,  if the -p option is present on the command line then the 4th field"
echo "            is interpreted  as the package ID and included in the file name"
echo "         5. FootPrint must be the next argument and is substituted for unknown in the template"
echo "         6. Optional symbol file modifier, if present, the value is used as a second suffix before"
echo "            the dash suffix. Normally this would be used to select the template with the correct pin"
echo "            -out, for example using ECB or BEC. The string is prefixed with a underscore composing"
echo "            the template file name."
echo "         7. Optional symbol file suffix, if this value is present in the parameter file, the"
echo "            string will be used between the template given on the commanline and the symbol"
echo "            file extension, in this case the name of the template given on the command line"
echo "            should only be the base file name. The string is prefixed with a dash composing"
echo "            the template file name. Note that if a modifier is not given then parameter in"
echo "            position 6 is assumed to be the symbol file suffix"
echo
echo "            Note on styles: The standard symbols include the technically correct symbols!"
echo "            Alternative styles can be used to provide a visual indication of something unusual"
echo "            about a component or group of components such as odd pin or package configuration"
echo "            When creating new styles a complementary symbol should also be created!"
echo
echo "Options: -h, --help       Show this information"
echo "         -v, --verbose    Display extra information during processing"
echo "             --version    Display software version information"
echo "         -p, --package    The packge type is included in the paratmeter data"
echo "         -t, --template   Include the template prefix and suffix in the output symbol name"
echo "         -m, --modifier   Suffix the template modifier in the output symbol name, like _ecb"
}
vecho()
{
   if [[ ! $1 = "" ]] ; then
     if $VERBOSE ; then
      echo $1
     fi
   fi
}

if [ "$1" = "-v" ] || [ "$1" = "--verbose" ] ; then VERBOSE=true ; shift ; fi

if [ "$1" = "-p" ] || [ "$1" = "--package" ] ; then include_package=true ; shift ; fi

do_Assimilate_Arguments(){

   for Arg in $*; do
      case $Arg in
         --help | -h) do_show_help ; exit 0                           ;;
        --usage | -u) do_show_usage ; exit 0                          ;;
      --verbose | -v) VERBOSE=true  ;                                 ;;
           --version) echo `basename $0` "Version $VER" ; exit 0      ;;
        --quiet | -q) QUIET=true  ;                                   ;;

      --package | -p) include_package=true ;                          ;;
     --template | -t) include_template=true ;                         ;;
     --modifier | -m) include_modifier=true ;                         ;;
                   *)                                                 ;;
      esac
   done

   return 0;
}

do_make_symbols(){
   vecho "Using Parameter File: $ParameterFile"
   while read -u3 line
   do
      parameters=( $line )
      PartNumber=${parameters[0]}
      Voltage=${parameters[1]}
      Features=${parameters[2]}
      if $include_package ; then
        Package=${parameters[3]}
        FootPrint=${parameters[4]}
        Template_file_modifier=${parameters[5]}
        Template_file_suffix=${parameters[6]}
      else
        Package=""
        FootPrint=${parameters[3]}
        Template_file_modifier=${parameters[4]}
        Template_file_suffix=${parameters[5]}
      fi

      vecho "Read PartNumber=$PartNumber, Voltage=$Voltage, Features=$Features, FootPrint=$FootPrint Template suffix1=$Template_file_modifier Template suffix2=$Template_file_suffix"

      # If not a comment and not a blank line
      if [ ! "${PartNumber:0:1}" = "#" ] && [ ! "$PartNumber" = "" ] ; then

         Value="$Voltage, $Features"

         FileNamePrefix=$PartNumber
         # IF not including we don't want underscores in the file name
         if $include_template; then
           FileNamePrefix=$FileNamePrefix"_"$Prefix
         fi

         FileNamePrefix=$FileNamePrefix"_"$Voltage"_"$Features

         if $include_package; then
            FileNamePrefix=$FileNamePrefix"_"$Package
         fi

         if [ "$Template_file_modifier" = "" ] ; then
           OutputFile=$FileNamePrefix$Extension
           cat $TemplateFile | sed -e s/"PART_NUMBER"/"$PartNumber"/g >$OutputFile.tp1
           cat $OutputFile.tp1 | sed -e s/"unknown"/"$FootPrint"/g >$OutputFile.tp2
           cat $OutputFile.tp2 | sed -e s/"COMPONENT_VALUE"/"$Value"/g >$OutputFile
           vecho "Mode 1: Created symbol:$OutputFile"
         else
           OutputFile=$FileNamePrefix
           if [ ! "$Template_file_suffix" = "" ] ; then
             TmpFileName=$TemplateFile"-"$Template_file_modifier"-"$Template_file_suffix$Extension
             if $include_modifier ; then
               OutputFile=$OutputFile"_"$Template_file_modifier"-"$Template_file_suffix
             else
               OutputFile=$OutputFile"-"$Template_file_suffix
             fi
           else
             TmpFileName=$TemplateFile"-"$Template_file_modifier$Extension
             if $include_modifier; then
               OutputFile=$FileNamePrefix"-"$Template_file_modifier
             fi
           fi

           OutputFile=$OutputFile$Extension

           if [ -f $TmpFileName ] ; then
             cat $TmpFileName    | sed -e s/"PART_NUMBER"/"$PartNumber"/g >$OutputFile.tp1
             cat $OutputFile.tp1 | sed -e s/"unknown"/"$FootPrint"/g >$OutputFile.tp2
             cat $OutputFile.tp2 | sed -e s/"COMPONENT_VALUE"/"$Value"/g >$OutputFile
             vecho "Mode 2:Created symbol:$OutputFile"
           else
              echo "Did not find template file:$TmpFileName"
              #could / should check for scond templates here because maybe we are suppose to
              # make these and not the first set
           fi
         fi
         # ------ Remove the tmp files
         if [ -f $OutputFile.tp1 ]; then
           rm -f $OutputFile*.tp1
         fi
         if [ -f $OutputFile.tp2 ]; then
           rm -f $OutputFile*.tp2
         fi
      fi
   done 3< <(cat $ParameterFile)
}

do_Reconciliation() {

  # clean command-line based on set parameters
  if $VERBOSE ; then shift ;fi
  if $QUIET ; then shift ;fi
  if $include_package ; then shift ;fi
  if $include_template ; then shift ;fi
  if $include_modifier ; then shift ;fi

  # Check the Template Arugment
  if [ -f $1 ] ; then
    TemplateFile="$1"
  else
    if [ -f "$1.sym" ] ; then
      TemplateFile="$1.sym"
      vecho "Using Template File: $TemplateFile"
    else
      files=$(ls $1*.sym 2>/dev/null)
      if [ ! -z "$files" ]; then
        # Assume for now that is the is symbol file exist
        TemplateFile="$1"
        vecho "Using $TemplateFile for the symbol template file name"
      fi
    fi
  fi

  if [ ! "$TemplateFile" = "" ] ; then
    shift
    Prefix=`echo "${TemplateFile}" | cut -d'-' -f1` ; # trim away all but first word
  else
    echo "Template File not found: $TemplateFile"
    do_show_usage
    return $ERR_BAD_ARGS
   fi

  vecho "Trimmed $Prefix from the symbol name to use in the output file name"

  if [ -f $1 ] ; then
    ParameterFile="$1"
    shift ;
  else
    if [  -f "$1.spm" ] ; then
      ParameterFile="$1.spm"
    else
      echo "Parameter File not found: $ParameterFile"
      do_show_usage
      return $ERR_BAD_ARGS
    fi
  fi

  return 0;
}

# =-=-=-=-=-=-=-=-=-=-=-=-=-=-> MAIN <-=-=-=-=-=-=-=-=-=-=-=-=-=-=-

# ----------------- Check for Empty Command Line ------------------
if [ "$#" -lt $MINARGS ] ; then do_show_usage; exit $ERR_BAD_ARGS ; fi

# ----------------------- Parse Command Line ----------------------
do_Assimilate_Arguments $*

if [ $? != 0 ]; then exit $?; fi

do_Reconciliation $*

if [ $? != 0 ]; then exit $?; fi

do_make_symbols


















